### PR TITLE
Use Pigweed's new Python build system

### PR DESCRIPTION
**Pull request by AnthonyDiGirolamo**
#### Problem
Updating Matter to use Pigweed's new Python build system. This helps scale bootstrap and build times for Python packages by not requiring any pip installations. Pip installations into the users `.environment/pigweed-venv` are now optional with the ability to merge multiple Python packages into a single package.

#### Change overview
This change reduces the number of pip installs performed during bootstrap and builds to just a single `pigweed` package along with any others specified in `//BUILD.gn`.

All python_actions performed duing a GN build do not rely on pip installs of any in-tree python packages including Pigweed ones. Instead, build a venv is created in the out directory with only 3rd party deps installed. All Python imports of in-tree packages are accomplished by setting PYTHONPATH for each python_action. This allows scaling up the number of in-tree Python packages without a significant impact to build speed.

Bootstrap time is slightly reduced however all Pigweed Python modules are installed along with the Python packages from:
- `//examples/common/pigweed/rpc_console`
- `//examples/chef`
- `//integrations/mobly`

Before:
```
  Setting up Python environment.....done (1m34.7s)
```

After:
```
  Setting up Python environment.....done (1m14.7s)
```

#### Testing
- Ran bootstrap from scratch and verified all python packages are available as expected. Also ran a default build:
```
gn gen out
ninja -C out python_packages default check
```
- Mostly relying on CI here since I can only manually test on Linux.

### CODE REVIEW

**Comment by AnthonyDiGirolamo** [comment]
Every test seems to be failing in bootstrap on the Python environment setup stage. This is probably due to pip compiling native extensions for 3rd party packages that aren't already in the pip-cache. On my local Linux workstation that stage takes about 1 min if nothing needs to be compiled and 5 min otherwise.

Running `pip cache purge` followed by `. ./scripts/bootstrap.sh` I see this in the logs `.environment/pigweed-venv/pip-requirements.log`:
```
2022-07-27T13:18:39,083 Building wheels for collected packages: construct, future, reedsolo, bitstring, dbus-python, pgi, mobly, cbor, cbor2, psutil, stringcase, tornado, cffi, gevent, pyyaml, timeout_decorator, docopt
...
2022-07-27T13:21:05,177 Successfully built construct future reedsolo bitstring dbus-python pgi mobly cbor cbor2 psutil stringcase tornado cffi gevent pyyaml timeout_decorator docopt
```

That stage dosen't happen when running `. ./scripts/bootstrap.sh` a second time. Those same deps are installed by default before my change and the versions are identical between master and my branch. Is there any pip-cache behavior in the the github actions?

**Comment by bzbarsky-apple** [comment]
> Is there any pip-cache behavior in the the github actions?

You mean across jobs?  No.  Each job starts with a blank slate of whatever is in the docker file.

Could we prime the pip cache in the docker file somehow?

**Comment by AnthonyDiGirolamo** [comment]
@bzbarsky-apple I think yes by adding something like `pip install --requirement scripts/requirements.txt --constraint scripts/constraints.txt` to the docker setups. Maybe here? 
https://github.com/project-chip/connectedhomeip/blob/master/integrations/docker/images/chip-build/Dockerfile#L91 
I'm not sure how to build a new docker image though and would need some guidance there.

Another option would be to pre-download/build all required third party wheels into a local directory before `. ./scripts/bootstrap.sh` is run. I wrote this and it seems to do the trick locally for me:

```
#!/usr/bin/bash
WHEEL_DIR=$(readlink -f "${PW_PROJECT_ROOT}/pip-wheels")
WHEEL_DOWNLOAD_DIR="${WHEEL_DIR}/download"

mkdir -p "${WHEEL_DIR}"
cd "${WHEEL_DIR}"

python3 -m venv venv
. ./venv/bin/activate

# Install the latest pip and setuptools
pip install --upgrade pip setuptools wheel toml

# Download and build wheels to save in the 'download' folder:
pip wheel --wheel-dir download pip setuptools wheel toml

# Pigweed's pinned deps
pip wheel --wheel-dir download \
  -r ../third_party/pigweed/repo/pw_env_setup/py/pw_env_setup/virtualenv_setup/constraint.list
# Matter's Packages
pip wheel --wheel-dir download \
  -r ../scripts/requirements.txt -c ../scripts/constraints.txt

PIP_CONF_FILE="${WHEEL_DIR}/pip.conf"

tee <<EOF > "${PIP_CONF_FILE}"
[global]
no-index = True
find-links = file://${WHEEL_DOWNLOAD_DIR}
EOF

echo "Set this environment variable to use the pip-wheels dir:"
echo "export PIP_CONFIG_FILE=${PIP_CONFIG_FILE}"
```

It downloads all wheels in to a `pip-wheels/download` folder. Then before running bootstrap I set the `PIP_CONFIG_FILE` env var to tell pip to get it's wheels from the local directory instead of downloading from pypi.org.

**Comment by bzbarsky-apple** [comment]
@woody-apple or @andy31415 (on vacation now) can rebuild the docker images.  A good start is creating the PR for the docker file and bumping the version in https://github.com/project-chip/connectedhomeip/blob/master/integrations/docker/images/chip-build/version.  Once that merges we can update the docker images, move the CI jobs to the new version, and then move forward....

**Comment by AnthonyDiGirolamo** [comment]
> @woody-apple or @andy31415 (on vacation now) can rebuild the docker images. A good start is creating the PR for the docker file

Thanks! I'll take a look at making a PR for that.

**Comment by AnthonyDiGirolamo** [comment]
This change is looking good now. There's one lingering failure that seems unrelated. https://github.com/project-chip/connectedhomeip/runs/7571087533?check_suite_focus=true

```
2022-07-29 01:01:11 INFO    clang-14: error: no such file or directory: '../../examples/all-clusters-app/linux/.environment/cipd/packages/pigweed/lib/libc++.a'
```

Was this failing before?

**Comment by mspang** [comment]
> This change is looking good now. There's one lingering failure that seems unrelated. https://github.com/project-chip/connectedhomeip/runs/7571087533?check_suite_focus=true
> 
> ```
> 2022-07-29 01:01:11 INFO    clang-14: error: no such file or directory: '../../examples/all-clusters-app/linux/.environment/cipd/packages/pigweed/lib/libc++.a'
> ```
> 
> Was this failing before?

Probably not, as our CI would be red.

Do we still need to bump the bootstrap times? This is breaking restyle for you, and it doesn't really make sense since this is supposed to be a perf improvement.

**Comment by mspang** [comment]
> > This change is looking good now. There's one lingering failure that seems unrelated. https://github.com/project-chip/connectedhomeip/runs/7571087533?check_suite_focus=true
> > ```
> > 2022-07-29 01:01:11 INFO    clang-14: error: no such file or directory: '../../examples/all-clusters-app/linux/.environment/cipd/packages/pigweed/lib/libc++.a'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Was this failing before?
> 
> Probably not, as our CI would be red.

Unrelated to this change, and will be solved by https://github.com/project-chip/connectedhomeip/pull/21437

**Comment by AnthonyDiGirolamo** [comment]
> Unrelated to this change, and will be solved by #21437

Nice thanks!

> Do we still need to bump the bootstrap times? 

Probably not. I increased them initially to get logs from CI. Will try reverting those changes after #21437 lands.

**Comment by woody-apple** [comment]
@AnthonyDiGirolamo thanks for your contribution! Can you please resolve the repo conflict, and I'd be happy to merge/publish as needed!

File: examples/common/pigweed/rpc_console/BUILD.gn
Comment (mspang) [inline examples/common/pigweed/rpc_console/BUILD.gn:31]: This sounds nondeterministic, can we avoid that please?

Comment (AnthonyDiGirolamo) [reply examples/common/pigweed/rpc_console/BUILD.gn:31]: `append_date_to_version = true` will be needed if this is installed as part of bootstrap. Reason is if there are python code changes pip will think it doesn't need to re-install this package unless the version is bumped. 

The alternative is to install `chip_rpc` as an editable python package. I've got that set now so I'll remove the append_date_to_version option here.

**Comment by AnthonyDiGirolamo** [comment]
Any suggestions on what the `ZAP / ZAP templates generation` error is about? It doesn't seem related to this change at first glance.

**Review by mspang — APPROVED** [review]

**Comment by AnthonyDiGirolamo** [comment]
@woody-apple Everything is good to go on this PR. Please let me know if I should change anything.

**Review by rgoliver — APPROVED** [review]

**Comment by AnthonyDiGirolamo** [comment]
> @woody-apple or @andy31415 (on vacation now) can rebuild the docker images. A good start is creating the PR for the docker file and bumping the version in https://github.com/project-chip/connectedhomeip/blob/master/integrations/docker/images/chip-build/version. Once that merges we can update the docker images, move the CI jobs to the new version, and then move forward....

@bzbarsky-apple Turns out this wasn't required. I updated just the pip packages that were missing Python v3.10 wheels on pypi.org. This PR is good to go, please take a look if you have time :smile:

**Review by bzbarsky-apple — APPROVED** [review]

**Pull request merged**
